docs/development/pigs/pig-016.rst

```diff
@@ -0,0 +1,215 @@
+image, cube, astro, time. Here we show the structure of the Gammapy package and
+consider possible changes to improve it.
+
+For Gammapy v0.14 in September 2019 (and same for v1.0), we propose the following changes:
+
+- move gammapy.utils.fitting to gammapy.modeling, and to move all models from
+  spectrum, image, cube and time to gammapy.modeling.models
+- dissolve gammapy.background, moving it's functionality into gammapy.spectrum
+  and gammapy.cube
+- dissolve gammapy.image, moving the useful functionality to gammapy.maps and
+  gammapy.cube.
+- Clean up gammapy.utils sub-packages
+
+After Gammapy v1.0, we propose to continue discussion on other changes
+considered while writing this PIG, described in the outlook section.
+
+Status
+======
+
+The current Gammapy package structure (as of v0.13):
+
+.. image:: pig-016-gammapy-package-organisation-status.png
+   :alt: Gammapy package structure status (v0.13)
+
+This illustrates the namespaces where the Gammapy API is exposed. The number of
+Python files is much larger (about 100) and not important for the end-user API.
+
+Proposal
+========
+
+We propose to introduce gammapy.modeling and gammapy.modeling.models, to
+dissolve gammapy.background and gammapy.image in the v0.14 development cycle,
+resulting in the following package structure for Gammapy v0.14 and v1.0:
+
+.. image:: pig-016-gammapy-package-organisation-proposal.png
+   :alt: Gammapy package structure proposal (v0.14)
+
+The following sections explain and motivate the proposed changes.
+
+Introduce gammapy.modeling
+--------------------------
+
+Currently our fitting and modeling framework is in gammapy.utils.fitting,
+and the built-in models in Gammapy are distributed like this:
+
+- gammapy.spectrum.models -- 1D spectral models
+- gammapy.image.models -- 2D spatial models
+- gammapy.cube.models -- 3D cube models
+- gammapy.time.models -- 1D time models
+
+Our proposal is to create a new sub-package gammapy.modeling and
+gammapy.modeling.models following the organisation chosen for models in Astropy and
+restructure code like this:
+
+- gammapy/utils/fitting -> gammapy/modeling/fitting
+- gammapy/utils/serialisation -> gammapy/modeling/serialisation
+- gammapy/spectrum/models -- gammapy/modeling/models/spectrum
+- gammapy/image/models -- gammapy/modeling/models/image
+- gammapy/cube/models -- gammapy/modeling/models/cube
+- gammapy/time/models -- gammapy/modeling/models/time
+
+Since it is very inconvenient for users to have to type sub-sub-sub-package
+names, we propose to expose the end-user API direction in gammapy.modeling for
+general classes (Parameter, Parameters, Model, Fit, Dataset). For the models, we
+propose to expose them in gammapy.modeling.models (same as Astropy), and to
+introduce a factory function to make it easy to make models that is easy to
+import and use. E.g. it could be gammapy.modeling.make_model and take a model
+specification string or dictionary.
```

Comment:
I would add a small code example of how this would work

Reply:
I don't want to add a code example / prototype implementation of `make_model` here in the PIG. It's not something that can be done in a few lines, but I think it's clear that it can be done (that's what e.g. astropy-regions does from DS9 region strings, or what ctools does from model XML files, or what SkyCoord does from all kinds of strings that describe coordinates, or many other examples exists - this involves a parser and an object mapper.

Reply:
Sorry for the confusion, I meant some example usages. I can see something like:
`pwl = make_model("powerlaw")`
But what if you want two? How do you do combined spectral and spatial?

Reply:
@registerrier - Designing the model input specification (a "domain specific language") is not in scope here. What would we gain if I guess at something here? How does this affect this proposal? That will require a lot of discussion and falls into the model serialisation work.

I think what bothers you (rightly so) are the long model class names? Specifically you don't like that I propose we add "Spectral" to all spectral models to make this uniform?
https://docs.gammapy.org/0.12/spectrum/index.html#id4

Or do you not believe we will have a model factory function, and thus don't like that I claim here we can develop it?

Reply:
I've added this guesswork for now:
```
Developing the syntax and schema for this
"model specification language" isn't in scope for this PIG. It will be done as
part of the effort on model serialisation in Gammapy in the coming months.
One could imagine that it uses YAML syntax like this::

    model = make_model("""
        type: spec, shape: powerlaw, pars: {norm:1e-12, index=2.7}
    """)

or a domain-specific language like this (a bit equivalent ot e.g. DS9 regions or
sky coordinates)::

    model = make_model("""
        spec.powerlaw(norm=1e-12, index=2.7)
    """)

or that one still uses Python syntax (not strings), only creates objects based
on attribute access, so calling it would look like this::

    model = Model.make.spec.powerlaw(norm="1e-12 cm-2 s-1 TeV", index=2.7)
```
Would prefer to remove it and only keep the sentence that it's not in scope to design this for this PIG, to avoid PIG review / discussion focusing on this, IMO mostly unrelated topic.
@registerrier @adonath  - Thoughts?

Reply:
I agree with @cdeil that proposing a solution for the model factory functions is not in scope for this PIG. I would not put any code at all. Just mention, that factory functions can be a solution to simplify imports of models for users.

Reply:
OK. Right indeed this would drive the PIG too far off. At least it is good for me to see some examples or what you had in mind. OK not to show them, but keep what is before 'One could imagine...'

This is likely a long process, and won't be in for 1.0. Right? 
Then I would add that 'in the coming months, after v1.0 release'

Reply:
That's what I put, and I added the explicit statement that it's not in scope and thus I'm not giving examples here, referencing model serialisation.

Reply:
I don't want to guess as to what's in v1.0 or not here. I would hope that we have a first version of YAML serialisation in v0.14 and a better one in v0.15 and we would start using it also as input model spec language.